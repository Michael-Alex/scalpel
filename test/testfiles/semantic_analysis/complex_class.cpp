namespace n
{
	namespace n2
	{
		class base1
		{
		};
	}
}

class base2
{
};

class base3
{
};

class test: public n::n2::base1, virtual base2, protected virtual base3
{
	public:
		test(const test& t);

		explicit
		test(int i);

		virtual
		~test();

		void*
		operator new[](long unsigned int, void*);

		void
		operator delete[](void*);

		const test&
		operator=(const test& t);

		bool
		operator==(const test& t) const;

		operator int() const;

		virtual
		void
		pure_virtual() = 0;

		static
		bool
		static_function();

	protected:
		void
		update()
		{
		}

	public:
		inline
		int
		value() const;

	private:
		static const bool b = false;
		int value_;
};

test::test(int)
{
}

test::test(const test&)
{
}

test::~test()
{
}

const test&
test::operator=(const test& rhs)
{
	value_ = rhs.value_;
	return *this;
}

bool
test::operator==(const test& rhs) const
{
	return value_ == rhs.value_;
}

void*
test::operator new[](long unsigned int, void*)
{
	return 0;
}

void
test::operator delete[](void*)
{
}

test::operator int() const
{
	return value_;
}

bool
test::static_function()
{
	return b;
}

int
test::value() const
{
	return value_;
}

void
test::pure_virtual()
{
}

