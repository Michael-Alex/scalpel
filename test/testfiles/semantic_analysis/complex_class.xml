<namespace id="n0">
 <namespace name="n" id="n1">
  <namespace name="n2" id="n2">
   <class name="super_test" id="c0">
    <destructor access="public" defined="false">
    </destructor>
   </class>
  </namespace>
 </namespace>
 <class name="virtual_base" id="c1">
  <destructor access="public" defined="false">
  </destructor>
 </class>
 <class name="virtual_base2" id="c2">
  <destructor access="public" defined="false">
  </destructor>
 </class>
 <class name="test" id="c3">
  <base_classes>
   <base_class id="c0" access="public"/>
   <base_class id="c1" access="private" virtual="true"/>
   <base_class id="c2" access="protected" virtual="true"/>
  </base_classes>
  <constructor access="public" explicit="true">
   <parameters>
    <parameter name="i">
     <type>
      <fundamental_type type="int"/>
     </type>
    </parameter>
   </parameters>
  </constructor>
  <constructor access="public">
   <parameters>
    <parameter name="t">
     <type>
      <reference>
       <const>
        <class id="c3"/>
       </const>
      </reference>
     </type>
    </parameter>
   </parameters>
  </constructor>
  <destructor access="public" virtual="true">
  </destructor>
  <operator_function operator="=" access="public">
   <return_type>
    <reference>
     <const>
      <class id="c3"/>
     </const>
    </reference>
   </return_type>
   <parameters>
    <parameter name="t">
     <type>
      <reference>
       <const>
        <class id="c3"/>
       </const>
      </reference>
     </type>
    </parameter>
   </parameters>
  </operator_function>
  <operator_function operator="==" access="public" const="true">
   <return_type>
    <fundamental_type type="bool"/>
   </return_type>
   <parameters>
    <parameter name="t">
     <type>
      <reference>
       <const>
        <class id="c3"/>
       </const>
      </reference>
     </type>
    </parameter>
   </parameters>
  </operator_function>
  <operator_function operator="new[]" access="public">
   <return_type>
    <pointer>
     <fundamental_type type="void"/>
    </pointer>
   </return_type>
   <parameters>
    <parameter>
     <type>
      <fundamental_type type="unsigned long int"/>
     </type>
    </parameter>
    <parameter>
     <type>
      <pointer>
       <fundamental_type type="void"/>
      </pointer>
     </type>
    </parameter>
   </parameters>
  </operator_function>
  <operator_function operator="delete[]" access="public">
   <return_type>
    <fundamental_type type="void"/>
   </return_type>
   <parameters>
    <parameter>
     <type>
      <pointer>
       <fundamental_type type="void"/>
      </pointer>
     </type>
    </parameter>
   </parameters>
  </operator_function>
  <conversion_function access="public" const="true">
   <return_type>
    <fundamental_type type="int"/>
   </return_type>
  </conversion_function>
  <simple_function name="static_function" access="public" static="true">
   <return_type>
    <fundamental_type type="bool"/>
   </return_type>
  </simple_function>
  <simple_function name="value" access="public" inline="true" const="true">
   <return_type>
    <fundamental_type type="int"/>
   </return_type>
  </simple_function>
  <simple_function name="pure_virtual" access="public" virtual="true" pure="true">
   <return_type>
    <fundamental_type type="void"/>
   </return_type>
  </simple_function>
  <simple_function name="update" access="protected">
   <return_type>
    <fundamental_type type="void"/>
   </return_type>
  </simple_function>
  <variable name="value_" access="private">
   <type>
    <fundamental_type type="int"/>
   </type>
  </variable>
  <variable name="b" access="private" static="true">
   <type>
    <const>
     <fundamental_type type="bool"/>
    </const>
   </type>
  </variable>
 </class>
</namespace>
