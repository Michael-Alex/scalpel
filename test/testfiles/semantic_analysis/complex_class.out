<namespace id="0">
 <namespace id="1" name="n">
  <namespace id="2" name="n2">
   <class id="0" name="base1">
    <destructor access="public" defined="false">
    </destructor>
   </class>
  </namespace>
 </namespace>
 <class id="1" name="base2">
  <destructor access="public" defined="false">
  </destructor>
 </class>
 <class id="2" name="base3">
  <destructor access="public" defined="false">
  </destructor>
 </class>
 <class id="3" name="test">
  <base_classes>
   <base_class  class_id="0" access="public"/>
   <base_class  class_id="1" access="private" virtual="true"/>
   <base_class  class_id="2" access="protected" virtual="true"/>
  </base_classes>
  <constructor access="public">
   <parameters>
    <parameter name="t">
     <type>
      <reference>
       <cv_qualified_type const="true">
        <class id="3"/>
       </cv_qualified_type>
      </reference>
     </type>
    </parameter>
   </parameters>
  </constructor>
  <constructor access="public" explicit="true">
   <parameters>
    <parameter name="i">
     <type>
      <fundamental_type type="int"/>
     </type>
    </parameter>
   </parameters>
  </constructor>
  <destructor access="public" virtual="true">
  </destructor>
  <operator_member_function id="0" operator="new[]" access="public">
   <return_type>
    <pointer>
     <fundamental_type type="void"/>
    </pointer>
   </return_type>
   <parameters>
    <parameter>
     <type>
      <fundamental_type type="unsigned long int"/>
     </type>
    </parameter>
    <parameter>
     <type>
      <pointer>
       <fundamental_type type="void"/>
      </pointer>
     </type>
    </parameter>
   </parameters>
  </operator_member_function>
  <operator_member_function id="1" operator="delete[]" access="public">
   <return_type>
    <fundamental_type type="void"/>
   </return_type>
   <parameters>
    <parameter>
     <type>
      <pointer>
       <fundamental_type type="void"/>
      </pointer>
     </type>
    </parameter>
   </parameters>
  </operator_member_function>
  <operator_member_function id="2" operator="=" access="public">
   <return_type>
    <reference>
     <cv_qualified_type const="true">
      <class id="3"/>
     </cv_qualified_type>
    </reference>
   </return_type>
   <parameters>
    <parameter name="t">
     <type>
      <reference>
       <cv_qualified_type const="true">
        <class id="3"/>
       </cv_qualified_type>
      </reference>
     </type>
    </parameter>
   </parameters>
  </operator_member_function>
  <operator_member_function id="3" operator="==" access="public" const="true">
   <return_type>
    <fundamental_type type="bool"/>
   </return_type>
   <parameters>
    <parameter name="t">
     <type>
      <reference>
       <cv_qualified_type const="true">
        <class id="3"/>
       </cv_qualified_type>
      </reference>
     </type>
    </parameter>
   </parameters>
  </operator_member_function>
  <conversion_function id="0" access="public" const="true">
   <return_type>
    <fundamental_type type="int"/>
   </return_type>
  </conversion_function>
  <simple_member_function id="0" name="pure_virtual" access="public" virtual="true" pure="true">
   <return_type>
    <fundamental_type type="void"/>
   </return_type>
  </simple_member_function>
  <simple_member_function id="1" name="static_function" access="public" static="true">
   <return_type>
    <fundamental_type type="bool"/>
   </return_type>
  </simple_member_function>
  <simple_member_function id="2" name="update" access="protected" inline="true">
   <return_type>
    <fundamental_type type="void"/>
   </return_type>
  </simple_member_function>
  <simple_member_function id="3" name="value" access="public" inline="true" const="true">
   <return_type>
    <fundamental_type type="int"/>
   </return_type>
  </simple_member_function>
  <variable id="0" name="b" access="private" static="true">
   <type>
    <cv_qualified_type const="true">
     <fundamental_type type="bool"/>
    </cv_qualified_type>
   </type>
  </variable>
  <variable id="1" name="value_" access="private">
   <type>
    <fundamental_type type="int"/>
   </type>
  </variable>
 </class>
</namespace>
